src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+        conn: &Connection,
+        parent_table_name: &str,
+        id_column_name: &str,
+        table_name: &str,
+        base_path: &Path,
+    ) -> CargoResult<()> {
+        let _p = crate::util::profile::start(format!("update db for removed {table_name}"));
+        trace!("checking for db entries to remove from {table_name}");
+        let mut select_stmt = conn.prepare_cached(&format!(
+            "SELECT {table_name}.rowid, {parent_table_name}.name, {table_name}.name
+             FROM {parent_table_name}, {table_name}
+             WHERE {table_name}.{id_column_name} = {parent_table_name}.id",
+        ))?;
+        let mut delete_stmt =
+            conn.prepare_cached(&format!("DELETE FROM {table_name} WHERE rowid = ?1"))?;
+        let mut rows = select_stmt.query([])?;
+        while let Some(row) = rows.next()? {
+            let rowid: i64 = row.get_unwrap(0);
+            let id_name: String = row.get_unwrap(1);
+            let name: String = row.get_unwrap(2);
+            if !base_path.join(id_name).join(name).exists() {
+                delete_stmt.execute([rowid])?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Removes database entries for any files that are not on disk for the parent tables.
+    fn update_db_parent_for_removed_from_disk(
+        conn: &Connection,
+        parent_table_name: &str,
+        base_path: &Path,
+        child_base_paths: &[&Path],
+        delete_paths: &mut Vec<PathBuf>,
+    ) -> CargoResult<()> {
+        let _p = crate::util::profile::start(format!(
+            "update db parent for removed from disk {parent_table_name}"
+        ));
+        trace!("checking for db entries to remove from {parent_table_name}");
+        let mut select_stmt =
+            conn.prepare_cached(&format!("SELECT rowid, name FROM {parent_table_name}"))?;
+        let mut delete_stmt =
+            conn.prepare_cached(&format!("DELETE FROM {parent_table_name} WHERE rowid = ?1"))?;
+        let mut rows = select_stmt.query([])?;
+        while let Some(row) = rows.next()? {
+            let rowid: i64 = row.get_unwrap(0);
+            let id_name: String = row.get_unwrap(1);
+            if !base_path.join(&id_name).exists() {
+                delete_stmt.execute([rowid])?;
+                // Make sure any child data is also cleaned up.
+                for child_base in child_base_paths {
+                    let child_path = child_base.join(&id_name);
+                    if child_path.exists() {
+                        debug!("removing orphaned path {child_path:?}");
+                        delete_paths.push(child_path);
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Updates the database to add any `.crate` files that are currently
+    /// not tracked (such as when they are downloaded by an older version of
+    /// cargo).
+    fn populate_untracked_crate(conn: &Connection, base_path: &Path) -> CargoResult<()> {
```

Comment:
Note to self (and others): this is about the point where my eyes are glazing over and not giving things as much attention